File: python/924.py
from typing import List

class Solution:
    def minMalwareSpread(self, graph: List[List[int]], initial: List[int]) -> int:
        st = set(initial)
        vis = [False] * len(graph)
        def dfs(x: int) -> None:
            vis[x] = True
            nonlocal node_id, size
            size += 1
            if node_id != -2 and x in st:
                node_id = x if node_id == -1 else -2
            for y, conn in enumerate(graph[x]):
                if conn and not vis[y]:
                    dfs(y)
        ans = -1
        max_size = 0
        for x in initial:
            if vis[x]:
                continue
            node_id = -1
            size = 0
            dfs(x)
            if node_id >= 0 and (size > max_size or node_id < ans and size == max_size):
                ans = node_id
                max_size = size
        return min(initial) if ans < 0 else ans

File: go/924.go
package main

import "slices"

func minMalwareSpread(graph [][]int, initial []int) int {
	vis := make([]bool, len(graph))
	isInitial := make([]bool, len(graph))
	for _, x := range initial {
		isInitial[x] = true
	}
	var nodeId, size int
	var dfs func(x int)
	dfs = func(x int) {
		vis[x] = true
		size++
		if nodeId != -2 && isInitial[x] {
			if nodeId == -1 {
				nodeId = x
			} else {
				nodeId = -2
			}
		}
		for y, conn := range graph[x] {
			if conn == 1 && !vis[y] {
				dfs(y)
			}
		}
	}

	ans := -1
	maxSize := 0
	for _, x := range initial {
		if vis[x] {
			continue
		}
		nodeId = -1
		size = 0
		dfs(x)
		if nodeId >= 0 && (size > maxSize || size == maxSize && nodeId < ans) {
			ans = nodeId
			maxSize = size
		}
	}
	if ans < 0 {
		return slices.Min(initial)
	}
	return ans
}
